File: 128.longest-consecutive-sequence.js
/*
 * @lc app=leetcode id=128 lang=javascript
 *
 * [128] Longest Consecutive Sequence
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  class UnionFind {
    union(x, y) {
      const rootX = this.find(x);
      const rootY = this.find(y);
      if (rootX === rootY) return;
      this[rootX] = rootY;
    }
    find(x) {
      if (undefined === this[x]) this[x] = x;
      return this[x] === x ? x : this.find(this[x]);
    }
  }

  const ufo = new UnionFind();
  nums.forEach((num) => {
    ufo.find(num);
    if (undefined !== ufo[num - 1]) ufo.union(num, num - 1);
    if (undefined !== ufo[num + 1]) ufo.union(num, num + 1);
  });

  const rootCount = new Map();
  new Set(nums).forEach((num) => {
    const root = ufo.find(num);
    rootCount.set(root, (rootCount.get(root) ?? 0) + 1);
  });
  return Math.max(...rootCount.values(), 0);
};
// @lc code=end

File: solutions/128.longest-consecutive-sequence.rb
# @param {Integer[]} nums
# @return {Integer}
def longest_consecutive(nums)
  res = 0
  set = nums.to_set
  set.each do |n|
    unless set.include?(n-1) # 什么样的数才是一个连续序列的起点？答案是这个数的前一个数不存在于数组中
      seq_len = 0
      seq_len += 1 while set.include?(n+seq_len) # 什么时候是终点呢？答案是当前数num的后一个数nunm + 1不存在于数组中
      res = [res, seq_len].max
    end
  end
  res
end
